chore: rebalance embedded login messaging and add hybrid-use framing (#1327)
auth0/docs-v2: main/docs/authenticate/login/native-login.mdx

description: Learn about the different login options for native mobile applications.
title: Native Login
---

> Native mobile applications can use native or browser-based login flows—or combine them within the same app.

Reviewer: ```suggestion
Native mobile applications can use native login flows, browser-based login flows, or both within the same application.
```

In a **browser-based** login flow, the user is shown a web browser and redirected to the Auth0 login page for sign up or log in. For example: an iOS application opens a SafariViewController or an Android application opens a Custom Chrome Tab.

> With a **native** login flow, the user signs up, signs in, enrolls factors, or registers passkeys directly inside the app.

Reviewer: ```suggestion
With a **native** login flow, the user signs up, signs in, enrolls factors, or registers passkeys directly inside the application.
```

> Auth0 supports either approach, and many native apps mix the two—for example, using a browser-based redirect for primary sign-in and embedded [Passkey APIs](/docs/authenticate/database-connections/passkeys) or [Native Social Login](/docs/authenticate/identity-providers/social-identity-providers) for re-authentication, factor enrollment, or step-up authentication inside the app.

Reviewer: ```suggestion
Native applications and use a combination of both native and browser-based loginf lows. For example, you can us a browser-based redirect for primary sign-in and embedded [Passkey APIs](/docs/authenticate/database-connections/passkeys) or [Native Social Login](/docs/authenticate/identity-providers/social-identity-providers) for re-authentication, factor enrollment, or step-up authentication inside the application.
```

* **<Tooltip tip="Single Sign-On (SSO): Service that, after a user logs into one applicaton, automatically logs that user in to other applications." cta="View Glossary" href="/docs/glossary?term=SSO">SSO</Tooltip>:** Cross-app SSO across native apps requires shared session state. [Native to Web SSO](/docs/authenticate/single-sign-on/native-to-web) lets a native app share a session with a web application. Sharing tokens directly across native apps via a shared keychain is not aligned with the OAuth 2.0 specifications.
* **Implementation effort:** A custom credential form takes more time to implement than the system browser flow that ships with the Auth0 native SDKs.
* **Feature delivery:** New features that affect the login UI must be added to the app and shipped to users. Server-side capabilities (for example, [API-driven MFA](/docs/secure/multi-factor-authentication) or [attack protection](/docs/secure/attack-protection)) can be enabled without an app update.

> * **OAuth 2.0 best practices ([RFC 8252](https://tools.ietf.org/html/rfc8252)):** RFC 8252 recommends external user-agents for browser-mediated OAuth flows in native apps. Modern embedded paths—Passkey APIs and Native Social Login—use standards-compliant token exchanges that meet the same security goals while keeping the user inside the application.

Reviewer: ```suggestion
When you implement native embedded login with a custom credential form, keep the following considerations in mind.

* **Phishing/security:** An unauthorized party could decompile or intercept traffic to or from your application to get the <Tooltip tip="Client ID: Identification value given to your registered resource from Auth0." cta="View Glossary" href="/docs/glossary?term=Client+ID">Client ID</Tooltip> and authentication URL. With this information, the unauthorized party could create a rogue application, upload it to an application store, and use it to phish for usernames, passwords, and <Tooltip tip="Access Token: Authorization credential, in the form of an opaque string or JWT, used to access an API." cta="View Glossary" href="/docs/glossary?term=Access+Tokens">access aokens</Tooltip>.

    Mitigations include using [App Attestation](/docs/secure/sender-constraining/configure-sender-constraining), [DPoP](/docs/secure/sender-constraining/demonstrating-proof-of-possession-dpop), and Passkey APIs to bind tokens and credentials to your client.

* **<Tooltip tip="Single Sign-On (SSO): Service that, after a user logs into one applicaton, automatically logs that user in to other applications." cta="View Glossary" href="/docs/glossary?term=SSO">Single sign-on</Tooltip> (SSO):** Cross-application SSO across native apps requires shared session state. [Native to Web SSO](/docs/authenticate/single-sign-on/native-to-web) lets a native app share a session with a web application. Sharing tokens directly across native apps via a shared keychain is not aligned with the OAuth 2.0 specifications.

* **Implementation effort:** A custom credential form can take more time to implement than the system browser flow that ships with the Auth0 native SDKs. Similarly, when we release we features that affect the login UI, you must add them the app and ship the update to users.

* **OAuth 2.0 best practices ([RFC 8252](https://tools.ietf.org/html/rfc8252)):** RFC 8252 recommends external user-agents for browser-mediated OAuth flows in native apps.

Modern embedded paths, such as [Passkey APIs](/docs/authenticate/database-connections/passkeys) and [Native Social Login](/docs/authenticate/identity-providers/social-identity-providers), are protocol-compliant token exchanges and avoid most of these trade-offs.
```
